### PR TITLE
Fixes #9632 - Add Profiler Support for BQ Arrays of Structs

### LINKED ISSUE #9632: Profiler fails to profile the bigquery tables in case of partition type of table and field of Record Datatype
**Affected module**
Ingestion Framework

**Describe the bug**
Profiler Ingestion fails to compute metrics/profiler data in case of schema containing RECORD  datatype as mentioned here 
image (1) 
OM parses above schema as array or struct based on if it’s repeated or not

image (2)

now when we try to run profiler for these partitioned table with these fields, then computed query from OM fails on the bigquery side as it doesn’t unnest this field as array while firing to bigquery, now since this query fails we are not getting the profiling results for these tables which is like 90% of our datasets. This issue can be resolved if we use UNNEST function of bigquery for those array fields while firing it to the bigquery.

**To Reproduce**
1. Create bigquery partitioned type of table.
2. Add fields in the schema of datatype RECORD in bigquery.
image (1)
3. Run the profiler workflow and see the failed profiler logs / Find the created profiler query and run in the bigquery console directly and see the error in the bigquery console.

**Encountered exception Logs**

```
[2023-01-03 10:38:55] INFO     {Ingestion:database_service:206} - Scanned [yyy.bq.schedule_record_event]
[2023-01-03 10:38:55] DEBUG    {Ingestion:noop_secrets_manager:46} - Adding auth provider security config using noop secrets' manager
[2023-01-03 10:38:55] DEBUG    {OMetaAPI:server_mixin:86} - Validating client and server versions
[2023-01-03 10:38:55] DEBUG    {OMetaAPI:client:177} - URL http://localhost:8585/api/v1/version, method GET
[2023-01-03 10:38:55] DEBUG    {OMetaAPI:client:178} - Data {'headers': {'Content-type': 'application/json', 'Authorization': 'Bearer no_token'}, 'allow_redirects': False, 'params': None}
[2023-01-03 10:38:55] DEBUG    {OMetaAPI:client:177} - URL http://localhost:8585/api/v1/version, method GET
[2023-01-03 10:38:55] DEBUG    {OMetaAPI:client:178} - Data {'headers': {'Content-type': 'application/json', 'Authorization': 'Bearer no_token'}, 'allow_redirects': False, 'params': None}
[2023-01-03 10:38:55] DEBUG    {OMetaAPI:client:177} - URL http://localhost:8585/api/v1/databaseSchemas/95601612-6dc1-4586-a138-144d185a403a, method GET
[2023-01-03 10:38:55] DEBUG    {OMetaAPI:client:178} - Data {'headers': {'Content-type': 'application/json', 'Authorization': 'Bearer no_token'}, 'allow_redirects': False, 'params': None}
[2023-01-03 10:38:58] INFO     {Profiler:core:376} - Computing profile metrics for yyy.bq.schedule_record_event...
[2023-01-03 10:39:04] DEBUG    {Profiler:stddev:83} - schedule_record_event.event_id has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:07] DEBUG    {Profiler:stddev:83} - schedule_record_event.event_source has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:11] DEBUG    {Profiler:stddev:83} - schedule_record_event.event_type has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:14] DEBUG    {Profiler:min_length:50} - Don't know how to process type FLOAT when computing MIN_LENGTH
[2023-01-03 10:39:14] DEBUG    {Profiler:max_length:50} - Don't know how to process type FLOAT when computing MAX_LENGTH
[2023-01-03 10:39:18] DEBUG    {Profiler:mean:71} - Don't know how to process type TIMESTAMP when computing MEAN
[2023-01-03 10:39:18] DEBUG    {Profiler:min_length:50} - Don't know how to process type TIMESTAMP when computing MIN_LENGTH
[2023-01-03 10:39:18] DEBUG    {Profiler:max_length:50} - Don't know how to process type TIMESTAMP when computing MAX_LENGTH
[2023-01-03 10:39:18] DEBUG    {Profiler:stddev:83} - schedule_record_event.event_time has type TIMESTAMP, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:21] DEBUG    {Profiler:mean:71} - Don't know how to process type TIMESTAMP when computing MEAN
[2023-01-03 10:39:21] DEBUG    {Profiler:min_length:50} - Don't know how to process type TIMESTAMP when computing MIN_LENGTH
[2023-01-03 10:39:21] DEBUG    {Profiler:max_length:50} - Don't know how to process type TIMESTAMP when computing MAX_LENGTH
[2023-01-03 10:39:21] DEBUG    {Profiler:stddev:83} - schedule_record_event.last_update_ts has type TIMESTAMP, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:25] DEBUG    {Profiler:stddev:83} - schedule_record_event.correlation_id has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:28] DEBUG    {Profiler:stddev:83} - schedule_record_event.record_source has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:32] DEBUG    {Profiler:mean:71} - Don't know how to process type DATE when computing MEAN
[2023-01-03 10:39:32] DEBUG    {Profiler:min_length:50} - Don't know how to process type DATE when computing MIN_LENGTH
[2023-01-03 10:39:32] DEBUG    {Profiler:max_length:50} - Don't know how to process type DATE when computing MAX_LENGTH
[2023-01-03 10:39:32] DEBUG    {Profiler:stddev:83} - schedule_record_event.create_date has type DATE, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:35] DEBUG    {Profiler:mean:71} - Don't know how to process type TIMESTAMP when computing MEAN
[2023-01-03 10:39:35] DEBUG    {Profiler:min_length:50} - Don't know how to process type TIMESTAMP when computing MIN_LENGTH
[2023-01-03 10:39:35] DEBUG    {Profiler:max_length:50} - Don't know how to process type TIMESTAMP when computing MAX_LENGTH
[2023-01-03 10:39:35] DEBUG    {Profiler:stddev:83} - schedule_record_event.source_record_create_ts has type TIMESTAMP, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:38] DEBUG    {Profiler:stddev:83} - schedule_record_event.client_country_code has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:42] DEBUG    {Profiler:stddev:83} - schedule_record_event.client_code has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:45] DEBUG    {Profiler:min_length:50} - Don't know how to process type INTEGER when computing MIN_LENGTH
[2023-01-03 10:39:45] DEBUG    {Profiler:max_length:50} - Don't know how to process type INTEGER when computing MAX_LENGTH
[2023-01-03 10:39:49] DEBUG    {Profiler:stddev:83} - schedule_record_event.client_location_name has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:53] DEBUG    {Profiler:stddev:83} - schedule_record_event.schedule_category_code has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:39:56] DEBUG    {Profiler:min_length:50} - Don't know how to process type INTEGER when computing MIN_LENGTH
[2023-01-03 10:39:56] DEBUG    {Profiler:max_length:50} - Don't know how to process type INTEGER when computing MAX_LENGTH
[2023-01-03 10:40:00] DEBUG    {Profiler:min_length:50} - Don't know how to process type FLOAT when computing MIN_LENGTH
[2023-01-03 10:40:00] DEBUG    {Profiler:max_length:50} - Don't know how to process type FLOAT when computing MAX_LENGTH
[2023-01-03 10:40:03] DEBUG    {Profiler:mean:71} - Don't know how to process type DATE when computing MEAN
[2023-01-03 10:40:03] DEBUG    {Profiler:min_length:50} - Don't know how to process type DATE when computing MIN_LENGTH
[2023-01-03 10:40:03] DEBUG    {Profiler:max_length:50} - Don't know how to process type DATE when computing MAX_LENGTH
[2023-01-03 10:40:03] DEBUG    {Profiler:stddev:83} - schedule_record_event.shift_start_date has type DATE, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:06] DEBUG    {Profiler:stddev:83} - schedule_record_event.shift_start_time has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:10] DEBUG    {Profiler:mean:71} - Don't know how to process type DATE when computing MEAN
[2023-01-03 10:40:10] DEBUG    {Profiler:min_length:50} - Don't know how to process type DATE when computing MIN_LENGTH
[2023-01-03 10:40:10] DEBUG    {Profiler:max_length:50} - Don't know how to process type DATE when computing MAX_LENGTH
[2023-01-03 10:40:10] DEBUG    {Profiler:stddev:83} - schedule_record_event.shift_end_date has type DATE, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:14] DEBUG    {Profiler:stddev:83} - schedule_record_event.shift_end_time has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:17] DEBUG    {Profiler:min_length:50} - Don't know how to process type INTEGER when computing MIN_LENGTH
[2023-01-03 10:40:17] DEBUG    {Profiler:max_length:50} - Don't know how to process type INTEGER when computing MAX_LENGTH
[2023-01-03 10:40:20] DEBUG    {Profiler:stddev:83} - schedule_record_event.worker_name has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:23] DEBUG    {Profiler:min_length:50} - Don't know how to process type INTEGER when computing MIN_LENGTH
[2023-01-03 10:40:23] DEBUG    {Profiler:max_length:50} - Don't know how to process type INTEGER when computing MAX_LENGTH
[2023-01-03 10:40:26] DEBUG    {Profiler:stddev:83} - schedule_record_event.workgroup_name has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:29] DEBUG    {Profiler:min_length:50} - Don't know how to process type INTEGER when computing MIN_LENGTH
[2023-01-03 10:40:29] DEBUG    {Profiler:max_length:50} - Don't know how to process type INTEGER when computing MAX_LENGTH
[2023-01-03 10:40:32] DEBUG    {Profiler:stddev:83} - schedule_record_event.job_name has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:36] DEBUG    {Profiler:min_length:50} - Don't know how to process type INTEGER when computing MIN_LENGTH
[2023-01-03 10:40:36] DEBUG    {Profiler:max_length:50} - Don't know how to process type INTEGER when computing MAX_LENGTH
[2023-01-03 10:40:40] DEBUG    {Profiler:stddev:83} - schedule_record_event.labor_role_name has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:43] DEBUG    {Profiler:stddev:83} - schedule_record_event.shift_type_code has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:47] DEBUG    {Profiler:stddev:83} - schedule_record_event.schedule_available_to_worker_flag has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:50] DEBUG    {Profiler:stddev:83} - schedule_record_event.trace_metadata.key_value.key has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:40:51] DEBUG    {Profiler:stddev:83} - schedule_record_event.trace_metadata.key_value.value has type VARCHAR, which is not listed as quantifiable. We won't compute STDDEV for it.
[2023-01-03 10:41:10] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:12] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:14] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:18] DEBUG    {Profiler:median:53} - Don't know how to process type TIMESTAMP when computing Median
[2023-01-03 10:41:20] DEBUG    {Profiler:median:53} - Don't know how to process type TIMESTAMP when computing Median
[2023-01-03 10:41:22] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:24] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:26] DEBUG    {Profiler:median:53} - Don't know how to process type DATE when computing Median
[2023-01-03 10:41:28] DEBUG    {Profiler:median:53} - Don't know how to process type TIMESTAMP when computing Median
[2023-01-03 10:41:30] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:32] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:36] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:38] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:44] DEBUG    {Profiler:median:53} - Don't know how to process type DATE when computing Median
[2023-01-03 10:41:46] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:48] DEBUG    {Profiler:median:53} - Don't know how to process type DATE when computing Median
[2023-01-03 10:41:50] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:54] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:41:58] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:01] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:05] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:06] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:08] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:10] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:12] DEBUG    {Profiler:median:53} - Don't know how to process type ARRAY when computing Median
[2023-01-03 10:42:13] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:15] DEBUG    {Profiler:median:53} - Don't know how to process type VARCHAR when computing Median
[2023-01-03 10:42:17] DEBUG    {Profiler:workflow:426} - Traceback (most recent call last):
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 203, in _execute
    self._query_job.result()
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/job/query.py", line 1499, in result
    do_get_result()
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/api_core/retry.py", line 283, in retry_wrapped_func
    return retry_target(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/api_core/retry.py", line 190, in retry_target
    return target()
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/job/query.py", line 1489, in do_get_result
    super(QueryJob, self).result(retry=retry, timeout=timeout)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/job/base.py", line 728, in result
    return super(_AsyncJob, self).result(timeout=timeout, **kwargs)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/api_core/future/polling.py", line 137, in result
    raise self._exception
google.api_core.exceptions.BadRequest: 400 Cannot access field key on a value with type ARRAY<STRUCT<key STRING, value STRING>> at [4:2163]

Location: US
Job ID: db1e3933-650a-4cba-b29e-8cd77d69f31a


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 719, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/_helpers.py", line 494, in with_closed_check
    return method(self, *args, **kwargs)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 166, in execute
    self._execute(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 205, in _execute
    raise exceptions.DatabaseError(exc)
google.cloud.bigquery.dbapi.exceptions.DatabaseError: 400 Cannot access field key on a value with type ARRAY<STRUCT<key STRING, value STRING>> at [4:2163]

Location: US
Job ID: db1e3933-650a-4cba-b29e-8cd77d69f31a


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/orm_profiler/api/workflow.py", line 418, in execute
    profile: TableProfile = self.profiler_obj.process(
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/orm_profiler/profiler/core.py", line 379, in process
    self.compute_metrics()
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/orm_profiler/profiler/core.py", line 364, in compute_metrics
    self.profile_entity()
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/orm_profiler/profiler/core.py", line 355, in profile_entity
    profile_results = self.profiler_interface.get_all_metrics(
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/interfaces/sqa_interface.py", line 299, in get_all_metrics
    profile, column = future.result()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/interfaces/sqa_interface.py", line 267, in compute_metrics_in_thread
    row = compute_metrics_registry.registry[metric_type.value](
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/interfaces/sqa_interface.py", line 487, in get_query_metrics
    row = runner.select_first_from_query(metric_query)
  File "/Users/data-catalog-airflow-plugins/ingestion/src/metadata/orm_profiler/profiler/runner.py", line 95, in select_first_from_query
    return query.first()
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2810, in first
    return self.limit(1)._iter().first()
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2894, in _iter
    result = self.session.execute(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1689, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1614, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1481, in _execute_clauseelement
    ret = self._execute_context(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1845, in _execute_context
    self._handle_dbapi_exception(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2026, in _handle_dbapi_exception
    util.raise_(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 719, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/_helpers.py", line 494, in with_closed_check
    return method(self, *args, **kwargs)
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 166, in execute
    self._execute(
  File "/Users/Desktop/airflow-plugins/data-catalog-airflow-plugins/venv/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 205, in _execute
    raise exceptions.DatabaseError(exc)
sqlalchemy.exc.DatabaseError: (google.cloud.bigquery.dbapi.exceptions.DatabaseError) 400 Cannot access field key on a value with type ARRAY<STRUCT<key STRING, value STRING>> at [4:2163]

Location: US
Job ID: db1e3933-650a-4cba-b29e-8cd77d69f36a

[SQL: /* {"app": "OpenMetadata", "version": "0.12.0.0"} */
WITH `only_once` AS 
(SELECT count(`trace_metadata`.`key_value`.`value`) AS `count_1` 
FROM (SELECT `schedule_record_event`.`event_id` AS `event_id`, `schedule_record_event`.`event_source` AS `event_source`, `schedule_record_event`.`event_type` AS `event_type`, `schedule_record_event`.`event_specversion` AS `event_specversion`, `schedule_record_event`.`event_time` AS `event_time`, `schedule_record_event`.`last_update_ts` AS `last_update_ts`, `schedule_record_event`.`correlation_id` AS `correlation_id`, `schedule_record_event`.`record_source` AS `record_source`, `schedule_record_event`.`create_date` AS `create_date`, `schedule_record_event`.`source_record_create_ts` AS `source_record_create_ts`, `schedule_record_event`.`client_country_code` AS `client_country_code`, `schedule_record_event`.`client_code` AS `client_code`, `schedule_record_event`.`client_location_id` AS `client_location_id`, `schedule_record_event`.`client_location_name` AS `client_location_name`, `schedule_record_event`.`schedule_category_code` AS `schedule_category_code`, `schedule_record_event`.`shift_id` AS `shift_id`, `schedule_record_event`.`shift_net_hours` AS `shift_net_hours`, `schedule_record_event`.`shift_start_date` AS `shift_start_date`, `schedule_record_event`.`shift_start_time` AS `shift_start_time`, `schedule_record_event`.`shift_end_date` AS `shift_end_date`, `schedule_record_event`.`shift_end_time` AS `shift_end_time`, `schedule_record_event`.`client_worker_id` AS `client_worker_id`, `schedule_record_event`.`worker_name` AS `worker_name`, `schedule_record_event`.`workgroup_id` AS `workgroup_id`, `schedule_record_event`.`workgroup_name` AS `workgroup_name`, `schedule_record_event`.`job_id` AS `job_id`, `schedule_record_event`.`job_name` AS `job_name`, `schedule_record_event`.`labor_role_id` AS `labor_role_id`, `schedule_record_event`.`labor_role_name` AS `labor_role_name`, `schedule_record_event`.`shift_type_code` AS `shift_type_code`, `schedule_record_event`.`schedule_available_to_worker_flag` AS `schedule_available_to_worker_flag`, `schedule_record_event`.`trace_metadata` AS `trace_metadata`, `schedule_record_event`.`trace_metadata`.`key_value` AS `trace_metadata_key_value`, `schedule_record_event`.`trace_metadata`.`key_value`.`key` AS `trace_metadata_key_value_key`, `schedule_record_event`.`trace_metadata`.`key_value`.`value` AS `trace_metadata_key_value_value` 
FROM `cs_gwfm`.`schedule_record_event` 
WHERE `event_time` >= %(event_time_1:STRING)s AND `event_time` <= %(event_time_2:STRING)s) AS `anon_1` GROUP BY `trace_metadata`.`key_value`.`value` 
HAVING count(`trace_metadata`.`key_value`.`value`) = %(count_2:INT64)s)
 SELECT count(*) AS `uniqueCount` 
FROM `only_once`
 LIMIT %(param_1:INT64)s]
[parameters: {'event_time_1': '2022-12-04', 'event_time_2': '2023-01-04', 'count_2': 1, 'param_1': 1}]
(Background on this error at: https://sqlalche.me/e/14/4xp6)

```

**Expected behavior**
Expectation is to use unnest operator of bigquery to unnest array type of fields in case of partitioned tables and execute the profiling without failure.

**Version:**
 - OS: ios
 - Python version: python3.9
 - OpenMetadata version: 0.12.0.0
 - OpenMetadata Ingestion package version: [e.g. `openmetadata-ingestion[docker]==0.12.0.0`]

**Additional context**

### PR DESCRIPTION
**Pull request by TeddyCr**
### Describe your changes:

Fixes #9632

Implements the logic to handle (in Bigquery):
- Displaying arrays' struct fields as children in the metadata ingestion
Screenshot 2023-04-14 at 8 44 35 AM
- Processing arrays' struct field for the profiler and DQ
Screenshot 2023-04-14 at 8 49 57 AM

**IMPORTANT ⚠️**
For the profiler, regular `ARRAY` (i.e. arrays without structs) will still be ignored and the computation skipped for these column types. 

#
### Type of change:

- [x] Bug fix
- [ ] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [ ] I have commented on my code, particularly in hard-to-understand areas. 
- [ ] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

- [x] I have added a test that covers the exact scenario we are fixing. For complex issues, comment the issue number in the test for future reference.

### CODE REVIEW

File: ingestion/src/metadata/ingestion/source/database/bigquery/metadata.py [hunk 1]
```diff
@@ -107,7 +123,9 @@ def get_columns(bq_schema):
             "precision": field.precision,
             "scale": field.scale,
             "max_length": field.max_length,
-            "system_data_type": str(col_type),
+            "system_data_type": _array_sys_data_type_repr(col_type)
```
Comment (TeddyCr) [inline ingestion/src/metadata/ingestion/source/database/bigquery/metadata.py:126]: In BQ, `str(col_type)` for a column of type `ARRAY<STRUCT<STRING>>` will be shown as `ARRAY`. Resulting in `def _parse_datatype_string` to not consider the column as an array when checking (line 288) the following

```python
if data_type.startswith("array<"):
```

File: ingestion/src/metadata/ingestion/source/database/column_type_parser.py [hunk 2]
```diff
@@ -291,11 +291,18 @@ def _parse_datatype_string(
...
+            data_type_string = {
                 "dataType": "ARRAY",
                 "arrayDataType": arr_data_type,
                 "dataTypeDisplay": data_type,
             }
+            if arr_data_type == "STRUCT":
+                children = ColumnTypeParser._parse_struct_fields_string(
+                    data_type[6:-1][7:-1]
+                )["children"]
+                data_type_string["children"] = children
```
Comment (TeddyCr) [inline ingestion/src/metadata/ingestion/source/database/column_type_parser.py:304]: We'll get the children if we are working with an ARRAY of STRUCTS

File: ingestion/src/metadata/profiler/metrics/system/system.py [hunk 3]
```diff
@@ -140,12 +141,17 @@ def _(
...
-            if row_jobs.query_type == "INSERT":
-                rows_affected = row_jobs.dml_statistics.get("inserted_row_count")
-            if row_jobs.query_type == "DELETE":
-                rows_affected = row_jobs.dml_statistics.get("deleted_row_count")
-            if row_jobs.query_type == "UPDATE":
-                rows_affected = row_jobs.dml_statistics.get("updated_row_count")
+            try:
+                if row_jobs.query_type == "INSERT":
+                    rows_affected = row_jobs.dml_statistics.get("inserted_row_count")
+                if row_jobs.query_type == "DELETE":
```
Comment (ayush-shah) [inline ingestion/src/metadata/profiler/metrics/system/system.py:147]: Enum

File: ingestion/src/metadata/profiler/processor/runner.py [hunk 4]
```diff
@@ -92,7 +101,15 @@ def select_first_from_table(self, *entities, **kwargs):
...
-        query = self._build_query(*entities, **kwargs).select_from(self.table)
+        # this needs to be set before we call
+        # _self._build_query(*args, **kwarsg)
+        # as we pop some args.
+        column_like = ColumnLike.create(kwargs)
+        query = handle_array(
+            self._build_query(*entities, **kwargs),
+            column_like,  # type: ignore
+            self.table,
+        )
```
Comment (ayush-shah) [inline ingestion/src/metadata/profiler/processor/runner.py:112]: repeated here, let's use a common method.

File: ingestion/src/metadata/profiler/processor/runner.py [hunk 5]
```diff
@@ -107,7 +124,15 @@ def select_all_from_table(self, *entities, **kwargs):
...
```
Comment (ayush-shah) [inline ingestion/src/metadata/profiler/processor/runner.py:135]: Let's add this in common method as well

**Review by ayush-shah — CHANGES_REQUESTED** [review]
Hey Teddy, left few comments, Thanks.

File: ingestion/src/metadata/profiler/processor/handle_partition.py [hunk 6]
```diff
@@ -113,14 +115,24 @@ def handle_and_execute(_self, *args, **kwargs):
...
-                query_results = _self._build_query(*args, **kwargs).select_from(
-                    _self._sample if self.sampled else _self.table
+                # this needs to be set before we call
+                # _self._build_query(*args, **kwarsg)
+                # as we pop some args.
+                column_like = ColumnLike.create(kwargs)
+                query_results = handle_array(
+                    _self._build_query(*args, **kwargs),
+                    column_like,  # type: ignore
+                    _self._sample if self.sampled else _self.table,
```
[the remark this answers is not in the document]
Comment (TeddyCr) [reply ingestion/src/metadata/profiler/processor/handle_partition.py:135]: Not sure I understand. `_self._build_query` is already a method and `handle_array` is already a function.

File: ingestion/src/metadata/profiler/orm/converter.py [hunk 7]
```diff
@@ -139,31 +143,48 @@ def build_orm_col(
...
+        **kwargs,
     )
 
+    # pylint: disable=protected-access
+    if table_service_type == databaseService.DatabaseServiceType.BigQuery and is_array:
+        sqa_column._is_array = is_array
+        sqa_column._array_col = name.split(".")[0]
+    else:
+        sqa_column._is_array = False
+        sqa_column._array_col = None
```
[the remark this answers is not in the document]
Comment (TeddyCr) [reply ingestion/src/metadata/profiler/orm/converter.py:167]: I don't understand this comment.

[the remark this answers is not in the document]
Comment (ayush-shah) [reply ingestion/src/metadata/profiler/orm/converter.py]:
```suggestion
    sqa_column._is_array = False
    sqa_column._array_col = None
    # pylint: disable=protected-access
    if table_service_type == databaseService.DatabaseServiceType.BigQuery and is_array:
        sqa_column._is_array = is_array
        sqa_column._array_col = name.split(".")[0]
```

**Review by ayush-shah — APPROVED** [review]

**Pull request merged**